File: src/tests/ts-loss-gillespie.cpp
#include <cstdio>
#include <iostream>
#include <vector>

#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>

#include <thread>
#include <algorithm>

#include <gillespie-algorithm.hpp>

// A Gillespie algorithm simulation of tumour suppressor loss

using clonal_expansion::gillespie_ssa::times_to_final_vertices;
using clonal_expansion::gillespie_ssa::print_kaplan_meier;

int main(int argc, char* argv[]) {
    int runs_per_thr = 1e7; // default values
    int seed = 1;
    if (argc < 3) {
        printf("Call this program with\n ./tsgillespie seed runs \n");
        //printf("type should be 3 or 4\n");
        return 1;
    } // else:
    int num_thr = std::thread::hardware_concurrency() - 2;
    seed = atoi(argv[1]);
    runs_per_thr = atoi(argv[2]) / num_thr;
    //int type = atoi(argv[3]);

    // System coefficients:
    double rloh = 5e-7;
    double mu = 5e-8;

    clonal_expansion::Model model(5);
    model.m_migr[0][1] = mu;
    model.m_migr[0][2] = rloh;
    model.m_migr[1][3] = 0.5 * mu;
    model.m_migr[1][4] = 0.5 * rloh;
    model.m_migr[2][4] = 0.5 * mu;
    // birth and death rates:
    model.m_birth = {0, 0.05, 0.03, 0, 0};
    model.m_death = {0, 0, 0, 0, 0};
    model.m_initial_pops = {1e6, 0, 0, 0, 0};

    // We want the chance that the first tumour detected is of type "type".
    // We want to stop the simulations when the first tumour is detected: i.e.
    // of any type. And then we want the fraction of these that have the chosen
    // type.
    std::vector<int> final_vertices = {3, 4};

    // Run many simulations and store the results:
    std::vector<std::pair<double,int>> all_times;

    // run (num_thr * runs_per_thr) simulations and store the times in
    // all_times:
    {
        std::vector<std::vector<std::pair<double,int>>> times(num_thr);
        {
            std::vector<std::thread> simulations(num_thr);

            // Run some simulations:
            for (int i = 0; i < num_thr; ++i) {
                // Use a different seed for each simulation:
                simulations.at(i) = std::thread(
                                    times_to_final_vertices, 
                                    model, seed + i, runs_per_thr,
                                    final_vertices, 
                                    std::ref(times[i]));
            }

            for (int i = 0; i < num_thr; ++i) {
                simulations.at(i).join();
            }
        }

        // Flatten and store results:
        for (auto time : times) {
            for (auto t2 : time) {
                all_times.push_back(t2);
            }
        }
    }

    std::sort(all_times.begin(), all_times.end());
    size_t study_population = all_times.size();
    double time_max = all_times.back().first;

    // Print results for both types:
    for (auto type : final_vertices) {
        std::cout << "Type " << type << ":" << std::endl;

        std::vector<double> mutant_times;
        for (auto& result : all_times) {
            if (result.second == type) {
                mutant_times.push_back(result.first);
            }
        }

        // Guard against invalid access:
        if (mutant_times.size() < 1) {
            std::cout << "No results" << std::endl;
            return 2;
        }

        // Kaplan-Meier plot:
        std::cout << "age, p1, p2," << std::endl;
        print_kaplan_meier(time_max, mutant_times, study_population);

        std::cout << std::endl;
    }

    return 0;
}

File: src/python/Test/ts-loss-gillespie.py
import pyffwd
import numpy as np
import os
import threading

def times_to_final_vertices(model, seed, runs_per_thr,
                            final_vertices, result_container):
    rng = pyffwd.GSL_RNG(seed)
    for _ in range(runs_per_thr):
        result = pyffwd.first_passage_time_multiple(
            rng, model, final_vertices)
        result_container.append(result)

def print_kaplan_meier(time_max, mutant_times, study_population):
    pyffwd.print_kaplan_meier(time_max,
                              pyffwd.RealVector(mutant_times),
                              study_population)

if __name__ == "__main__":
      # default values
    runs_per_thr = 10
    seed = 1

    import sys
    if len(sys.argv) < 3:
        print("Call this program with\n ./tsgillespie seed runs \n")
        sys.exit(1)

    seed = int(sys.argv[1])
    total_runs = int(sys.argv[2])
    
    num_thr = max(1, os.cpu_count() - 2)  
    runs_per_thr = total_runs // num_thr

    # System coefficients:
    rloh = 5e-7
    mu = 5e-8

    # Define the model
    model = pyffwd.Model.create_model(
        5, 
        birth_rates=[0, 0.05, 0.03, 0, 0],
        death_rates=[0, 0, 0, 0, 0],
        initial_pops=[1e6, 0, 0, 0, 0]
    )
    model.m_migr = [
        {1: mu, 2: rloh},
        {3: 0.5 * mu, 4: 0.5 * rloh},
        {4: 0.5 * mu},
        {},
        {}
    ]

    final_vertices = pyffwd.IntVector([3, 4])

    all_times = []

    times = [[] for _ in range(num_thr)]
    threads = []

    for i in range(num_thr):
        thread = threading.Thread(
            target=times_to_final_vertices, 
            args=(model, seed + i, runs_per_thr,
                  final_vertices, times[i])
        )
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    for thread_results in times:
        all_times.extend(thread_results)

    all_times.sort(key=lambda x: x[0])

    study_population = len(all_times)
    time_max = all_times[-1][0] if all_times else 0

    # Print Kaplan-Meier
    for tumour_type in final_vertices:
        print(f"Type {tumour_type}:")

        mutant_times = [result[0] for result in all_times 
                        if result[1] == tumour_type]

        if not mutant_times:
            print("No results")
            sys.exit(2)

        # Print the Kaplan-Meier curve
        print("age, p1, p2,", flush=True)
        print_kaplan_meier(time_max, mutant_times, study_population)
        print()
